UNSPLIT AT statement page and diagrams

ericharmeling opened the pull request:
- Added new page for `UNSPLIT AT` statement.
- Added new diagrams for `UNSPLIT AT` statement.
- Added new diagrams/parameter for `SPLIT AT` statement for WITH EXPIRATION.
- Changed all `UNSPLIT AT` examples to use `movr` database.

Fixes #4895.
Fixes #5008.
Fixes #5007.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/docs/5360)
<!-- Reviewable:end -->

cockroach-teamcity commented:
Online preview: http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/ac0dc660f33721ab52a3ea7ae4cbf39a21911089/

Edited pages:
- [v19.2/alter-index.md](http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/ac0dc660f33721ab52a3ea7ae4cbf39a21911089/v19.2/alter-index.html)
- [v19.2/alter-table.md](http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/ac0dc660f33721ab52a3ea7ae4cbf39a21911089/v19.2/alter-table.html)
- [v19.2/split-at.md](http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/ac0dc660f33721ab52a3ea7ae4cbf39a21911089/v19.2/split-at.html)
- [v19.2/sql-statements.md](http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/ac0dc660f33721ab52a3ea7ae4cbf39a21911089/v19.2/sql-statements.html)
- [v19.2/unsplit-at.md](http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/ac0dc660f33721ab52a3ea7ae4cbf39a21911089/v19.2/unsplit-at.html)

rmloveland commented on _includes/v19.2/sql/diagrams/unsplit.html:
```diff
@@ -0,0 +1,25 @@
+<div><svg width="631" height="81">
```
This file does not appear to be included from any pages.  Unless I'm missing it?

rmloveland commented on v19.2/alter-index.md:
```diff
@@ -17,3 +17,4 @@ Subcommand | Description
 [`CONFIGURE ZONE`](configure-zone.html) | [Configure replication zones](configure-replication-zones.html) for an index.
 [`RENAME`](rename-index.html) | Change the name of an index.
 [`SPLIT AT`](split-at.html) | Force a key-value layer range split at the specified row in the index.
+[`UNSPLIT AT`](unsplit-at.html) | Remove a key-value layer range split enforcement at a specified row in the index.
```
I think we should add the "new in 19.2" flag to all of these new links to UNSPLIT AT, the same as we've done on the 'SQL Statements' page.

rmloveland commented on v19.2/alter-table.md:
```diff
@@ -35,6 +35,7 @@ Subcommand | Description | Can combine with other subcommands?
 [`RENAME CONSTRAINT`](rename-constraint.html) | Change constraints columns. | Yes
 [`RENAME TABLE`](rename-table.html) | Change the names of tables. | No
 [`SPLIT AT`](split-at.html) | Force a key-value layer range split at the specified row in the table. | No
+[`UNSPLIT AT`](unsplit-at.html) | Remove a key-value layer range split enforcement at a specified row in the table. | No
```
Same comment as previously re: "new in 19.2" flag.

rmloveland commented on v19.2/split-at.md:
```diff
@@ -252,3 +252,4 @@ SHOW RANGES FROM TABLE t;
 - [Distribution Layer](architecture/distribution-layer.html)
 - [Replication Layer](architecture/replication-layer.html)
 - [`SHOW JOBS`](show-jobs.html)
+- [`UNSPLIT AT`](unsplit-at.html)
```
Same comment as previously re: "new in 19.2" flag.  Also, I think the SPLIT AT page should be updated to link to the Range Merges doc and to note that you have to explicitly turn off Range Merges if you want to use SPLIT AT.  Perhaps that should be a separate issue?  UPDATE: I did a search and found that https://github.com/cockroachdb/docs/issues/4960 has been filed to address this.  I've added it to the 19.2 milestone as a P-1.

rmloveland commented on v19.2/unsplit-at.md:
```diff
@@ -0,0 +1,192 @@
+---
+title: UNSPLIT AT
+summary: The UNSPLIT AT statement reverts a manual key-value layer range split at a specified row in a table or index.
```
This summary could be updated to use the same "split enforcement" terminology as the doc.  I must admit I find "reverts a range split" more understandable than "removes a split enforcement", but I guess the latter is a more correct description of what happens?

rmloveland commented on v19.2/unsplit-at.md:
```diff
@@ -0,0 +1,192 @@
+toc: true
+---
+
+The `UNSPLIT AT` [statement](sql-statements.html) removes a [split enforcement](split-at.html) on a key-value layer range [split](architecture/distribution-layer.html#range-splits), at a specified row in a table or index.
```
I would make "range split" the link text.  It would be nice to have another sentence or two that describes what happens when the "split enforcement" is removed, even if it's just something simple like "this allows CRDB to merge the ranges back together at its leisure" or similar (but better :-} ).  Additionally, is there some cluster setting that would control how slowly or quickly this merging will happen?  If so I think we should mention that and add linkage.  Also, I think we should add a callout that links to the [Range Merges doc](https://www.cockroachlabs.com/docs/v19.1/range-merges.html) and mentions that the range merge cluster setting.  Finally I think we should make the behavior of how this SQL statement interacts with that setting clear, e.g., "for the UNSPLIT AT statement's removal of the split enforcement to result in ranges being merged back together at the specified point, the kv.range_merge.queue_enabled setting must be set to true"  (If that sentence is correct, I have not verified/tested that)

rmloveland commented on v19.2/unsplit-at.md:
```diff
@@ -0,0 +1,192 @@
+
+<div>
+  {% include {{ page.version.version }}/sql/diagrams/unsplit_index_at.html %}
+</div>
```
These diagrams look great!

rmloveland commented on v19.2/unsplit-at.md:
```diff
@@ -0,0 +1,192 @@
+
+## Required privileges
+
+The user must have the `INSERT` [privilege](authorization.html#assign-privileges) on the table or index.
```
I wish we could link directly to this privilege somehow.

rmloveland commented on v19.2/unsplit-at.md:
```diff
@@ -0,0 +1,192 @@
+
+ Parameter | Description
+-----------|-------------
+ `table_name` | The name of the table that should be unsplit.
```
Same comment as above re: standardizing on whether we are saying "unsplit this" vs. "removing split enforcements".  If we stick with "remove split enforcement" I think we should describe how that works generally in the overview as noted in previous comment.

rmloveland commented on v19.2/unsplit-at.md:
```diff
@@ -0,0 +1,192 @@
+ Parameter | Description
+-----------|-------------
+ `table_name` | The name of the table that should be unsplit.
+ `table_name @ index_name`<br>`standalone_index_name` | The name of the index that should be unsplit.
```
Same as above re: "unsplit" vs. "remove split enforcement".

rmloveland commented on v19.2/unsplit-at.md:
```diff
@@ -0,0 +1,192 @@
+-----------|-------------
+ `table_name` | The name of the table that should be unsplit.
+ `table_name @ index_name`<br>`standalone_index_name` | The name of the index that should be unsplit.
+ `select_stmt` | A [selection query](selection-queries.html) that produces one or more rows at which to unsplit the table or index.
```
Same as above re: "unsplit" vs. "remove split enforcement".  Another option is to clarify in the intro that "unsplit" actually means remove split enforcement, but hereafter we will just say unsplit", which IMO would be a fine way to address this while simplifying the verbiage.

rmloveland commented on v19.2/unsplit-at.md:
```diff
@@ -0,0 +1,192 @@
+
+## Examples
+
+{% include {{page.version.version}}/sql/movr-statements-nodes.md %}
```
Could the text of `--nodes` and `--demo-locality` be made into links to the specific flags?  This can be accomplished by adding an anchor `<a name="nodes-flag"></a>` inside the table on the `cockroach demo` page.  Also, this says "the --nodes ... tags" but I think you meant "the --nodes ... *flags*" ??? Unless "tags" is an alternative name like "soda" vs. "pop" which I was not aware of.  But I would still vote "flag" for consistency with our other docs. Or "options" could work.

rmloveland commented on v19.2/unsplit-at.md:
```diff
@@ -0,0 +1,192 @@
+
+### Unsplit a table
+
+The `crdb_internal.ranges` table contains information about ranges in your CockroachDB cluster. At this point, just one range contains the data in the `users` table:
```
I would reword the last clause slightly to "At this point, the data in the `users` table can be stored in just one range".  The table is the more important "thing" to a SQL user while the range is merely there in service of it.

rmloveland commented on v19.2/unsplit-at.md:
```diff
@@ -0,0 +1,192 @@
+(4 rows)
+~~~
+
+The `split_enforced_until` column is now `NULL` for all ranges in the `users` table. The table is still split into ranges at `'chicago'`, `'new york'`, and `'seattle'`, but the split is no longer enforced.
```
Might be good to mention that "now the system will be free to re-merge the ranges at its convenience *once you re-enable the `kv.range_merge.queue_enabled` setting to `true`*.  At least according to [the Range Merges doc](https://www.cockroachlabs.com/docs/v19.1/range-merges.html) you will need to have turned off the range merge queue to even be able to SPLIT AT - if I am reading that correctly?  I think we should at least make mention of that in the overview (I'll edit the comment on the overview as well).

rmloveland commented on v19.2/unsplit-at.md:
```diff
@@ -0,0 +1,192 @@
+~~~
+~~~
+  range_id |     start_key     |   start_pretty   |      end_key      |    end_pretty    | database_name | table_name | index_name  | replicas | learner_replicas |       split_enforced_until       | lease_holder
++----------+-------------------+------------------+-------------------+------------------+---------------+------------+-------------+----------+------------------+----------------------------------+--------------+
```
Note that we have more updates to do re: `split_enforced_until` incoming via https://github.com/cockroachdb/cockroach/pull/38004 (namely a WITH EXPIRATION option to SPLIT AT) but that will happen via another ticket.

rmloveland commented on v19.2/unsplit-at.md:
```diff
@@ -0,0 +1,192 @@
+(5 rows)
+~~~
+
+Now unsplit the index:
```
After looking at this even longer I think we should use the "remove split enforcement, let the range merge queue run AKA unsplit" explanation in the intro and then just say "unsplit" everywhere else down-page.

rmloveland commented on v19.2/unsplit-at.md:
```diff
@@ -0,0 +1,192 @@
+(5 rows)
+~~~
+
+The `split_enforced_until` column is now `NULL` for all ranges in the `rides` table. The table is still split into ranges at `25.00`, `50.00`, and `75.00`, but the split is no longer enforced.
```
Same comment as above re: giving a little more color about when/how the system merges the ranges back together.

rmloveland commented on v19.2/unsplit-at.md:
```diff
@@ -0,0 +1,192 @@
+
+
+## See also
+
```
I think we should also link to [the Range Merges doc](https://www.cockroachlabs.com/docs/v19.1/range-merges.html) here.

cockroach-teamcity commented:
Online preview: http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/ea3fad740bfff40f42ce41eded69dc12458d82ef/

Edited pages:
- [v19.2/alter-index.md](http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/ea3fad740bfff40f42ce41eded69dc12458d82ef/v19.2/alter-index.html)
- [v19.2/alter-table.md](http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/ea3fad740bfff40f42ce41eded69dc12458d82ef/v19.2/alter-table.html)
- [v19.2/split-at.md](http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/ea3fad740bfff40f42ce41eded69dc12458d82ef/v19.2/split-at.html)
- [v19.2/sql-statements.md](http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/ea3fad740bfff40f42ce41eded69dc12458d82ef/v19.2/sql-statements.html)
- [v19.2/unsplit-at.md](http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/ea3fad740bfff40f42ce41eded69dc12458d82ef/v19.2/unsplit-at.html)

ericharmeling commented:
@jseldess Added you for a quick pass before merging.

cockroach-teamcity commented:
Online preview: http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/23196ce8170c7c66e4a829ea625b6fbd11438983/

Edited pages:
- [v19.2/alter-index.md](http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/23196ce8170c7c66e4a829ea625b6fbd11438983/v19.2/alter-index.html)
- [v19.2/alter-table.md](http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/23196ce8170c7c66e4a829ea625b6fbd11438983/v19.2/alter-table.html)
- [v19.2/split-at.md](http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/23196ce8170c7c66e4a829ea625b6fbd11438983/v19.2/split-at.html)
- [v19.2/sql-statements.md](http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/23196ce8170c7c66e4a829ea625b6fbd11438983/v19.2/sql-statements.html)
- [v19.2/unsplit-at.md](http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/23196ce8170c7c66e4a829ea625b6fbd11438983/v19.2/unsplit-at.html)

ericharmeling commented:
@nvanbenschoten @andreimatei Thank you for the feedback!

I do have a final question...

When I create an index, it appears that CRDB performs a manual split, but with a much shorter expiration time on the enforcement.

```
root@127.0.0.1:63845/movr> SELECT range_id, start_pretty, end_pretty, split_enforced_until FROM crdb_internal.ranges WHERE table_name='rides';
  range_id | start_pretty | end_pretty | split_enforced_until
+----------+--------------+------------+----------------------+
        23 | /Table/55    | /Table/56  | NULL
(1 row)

Time: 2.002ms

root@127.0.0.1:63845/movr> CREATE INDEX revenue_idx ON rides(revenue);
CREATE INDEX

Time: 91.527ms

root@127.0.0.1:63845/movr> SELECT range_id, start_pretty, end_pretty, split_enforced_until FROM crdb_internal.ranges WHERE table_name='rides';
  range_id | start_pretty | end_pretty  |      split_enforced_until
+----------+--------------+-------------+---------------------------------+
        23 | /Table/55    | /Table/55/4 | NULL
        27 | /Table/55/4  | /Table/56   | 2019-09-10 21:37:44.47542+00:00
(2 rows)

Time: 1.346ms
```

What's going on here? When we create indexes, does this temporary split always happen?

nvanbenschoten commented:
> What's going on here? When we create indexes, does this temporary split always happen?

I wasn't aware that it ever happened. @andreimatei do you know about this?

cockroach-teamcity commented:
Online preview: http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/aeccba88e88e8eb080f4a523dfcb7a29990e8ad5/

Edited pages:
- [v19.2/alter-index.md](http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/aeccba88e88e8eb080f4a523dfcb7a29990e8ad5/v19.2/alter-index.html)
- [v19.2/alter-table.md](http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/aeccba88e88e8eb080f4a523dfcb7a29990e8ad5/v19.2/alter-table.html)
- [v19.2/split-at.md](http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/aeccba88e88e8eb080f4a523dfcb7a29990e8ad5/v19.2/split-at.html)
- [v19.2/sql-statements.md](http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/aeccba88e88e8eb080f4a523dfcb7a29990e8ad5/v19.2/sql-statements.html)
- [v19.2/unsplit-at.md](http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/aeccba88e88e8eb080f4a523dfcb7a29990e8ad5/v19.2/unsplit-at.html)

ericharmeling commented:
@jseldess Hate to do this again, but can you re-review? I changed all of the examples for `SPLIT AT` to use `movr`, added an example for `WITH EXPIRATION`, and I updated the diagrams/parameters to reflect the `WITH EXPIRATION` change.

cockroach-teamcity commented:
Online preview: http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/e6b54983a3b7d53d38f6adab47bbcecb8c365d7c/

Edited pages:
- [v19.2/alter-index.md](http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/e6b54983a3b7d53d38f6adab47bbcecb8c365d7c/v19.2/alter-index.html)
- [v19.2/alter-table.md](http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/e6b54983a3b7d53d38f6adab47bbcecb8c365d7c/v19.2/alter-table.html)
- [v19.2/split-at.md](http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/e6b54983a3b7d53d38f6adab47bbcecb8c365d7c/v19.2/split-at.html)
- [v19.2/sql-statements.md](http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/e6b54983a3b7d53d38f6adab47bbcecb8c365d7c/v19.2/sql-statements.html)
- [v19.2/unsplit-at.md](http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/e6b54983a3b7d53d38f6adab47bbcecb8c365d7c/v19.2/unsplit-at.html)

jseldess commented:
___
*[v19.2/split-at.md, line 106 at r5](https://reviewable.io/reviews/cockroachdb/docs/5360#-LowyyYg5alPaST95fA8:-LpepBxP9b_AXHapi_EX:b-bfrcz2) ([raw file](https://github.com/cockroachdb/docs/blob/aeccba88e88e8eb080f4a523dfcb7a29990e8ad5/v19.2/split-at.md#L106)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

You're asking why we have manual splits at all? Isn't that discussed in the "Why manually split a range?" section? Manually splitting is certainly much less important these days than it used to be, now that we have load based splitting to automatically split hotspots based on load, but it's still an option we provide. Maybe I'm missing your question though.

With respect to the UUID discussion: UUIDs are actually pretty easy to split because we know their distribution - it's exactly uniformly distributed. So splitting at the midway point is something like `80000000-00000000-00000000-00000000`.
</blockquote></details>

Based on my conversation with Eric, I think the question is really: Why should we show users how to split based on the *second* value in a compound primary key. With partitioning, we partition (auto-split) based on the first value, typically the geo-column. But I don't understand when a user would ever want, or need, to split on the second value. Do you have insights?


<!-- Sent from Reviewable.io -->

ericharmeling commented:
Thanks @nvanbenschoten !

> You're asking why we have manual splits at all? Isn't that discussed in the "Why manually split a range?" section? Manually splitting is certainly much less important these days than it used to be, now that we have load based splitting to automatically split hotspots based on load, but it's still an option we provide. Maybe I'm missing your question though.

I meant "in the context of partitioning", in the section "Split a table with a composite primary key" - totally left that out of my question though. To resolve that, I'm just going to remove "(e.g., when working with partitions)" from that section.

cockroach-teamcity commented:
Online preview: http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/8bcf770f9d113c6f3d5f272ce989fed58c8031e7/

Edited pages:
- [v19.2/alter-index.md](http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/8bcf770f9d113c6f3d5f272ce989fed58c8031e7/v19.2/alter-index.html)
- [v19.2/alter-table.md](http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/8bcf770f9d113c6f3d5f272ce989fed58c8031e7/v19.2/alter-table.html)
- [v19.2/split-at.md](http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/8bcf770f9d113c6f3d5f272ce989fed58c8031e7/v19.2/split-at.html)
- [v19.2/sql-statements.md](http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/8bcf770f9d113c6f3d5f272ce989fed58c8031e7/v19.2/sql-statements.html)
- [v19.2/unsplit-at.md](http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/8bcf770f9d113c6f3d5f272ce989fed58c8031e7/v19.2/unsplit-at.html)
